tests/python/unittest/test_auto_scheduler_feature.py

```diff
@@ -220,12 +220,15 @@ def tir_matmul(
 
 
 def test_primfunc():
-    features = auto_scheduler.feature.named_features_from_primfunc(tir_matmul)
```

Comment:
Would be good to keep the original pass because that is the original intended usage, since the original function was designed for auto-scheduler with limited passes (to enable fast feature extraction). The test should cover the original intended use.

The additional support enables possible explorations for new use cases. Consider add another testcase test_primfunc_lowered_with_let to get the fully lowered version. 

It would be helpful to construct a simplified TVM script version of the lowered version to see exactly where let bound may results in a change of values.

Reply:
+1 for TQ's idea

Reply:
I wrote the original test and its intended usage was for analysis after lowering.

I specifically did not want to construct a test from the explicitly lowered tvmscript so future large changes like CSE would cause the test to fail.

Reply:
I see, explicitly constructed script would help to clarify the regression that the change intend to cover.  

To rephrase what you are saying (correct if that was not what you meant):
- In the future, if CSE changes its behavior to generate other cases, the feature extractor from an explicit tvm-script may not capture that changes (as new CSE will generate new IRs that may have other properties) -- the test won't fail but I guess you mean it fails to capture the possible problem in CSE changes.

This is indeed a fair pt. Although running things through is a bit like integration tests(that tests the combined effect of CSE, lowering and feature extraction).

The case of tvmscript inputs is closer to unit-test to test the intended behavior of the specific pass. Such unit-tests usually have a clear intend, and easier to debug when it triggers a regression. Although its coverage won't move as the integration tests do. 

I still would suggest  keep `test_primfunc` given the original intended behavior of the pass was limited. It does not hurt to have a case to cover the original intended case. 

Adding another `test_primfunc_lowered_with_let` would give quite clear context, and provide some safety net when we start to explore the ability of using the extractor with CSE passes. 

As for the unit-test case, it is merely a recommendation(as per above discussion on unit-test and integration tests)

Reply:
Great, seems my interpretation aligns with what you mean. I agree that having a test that thread through lowered is useful if we bring in future usecases that applies feature extractor on lowered code.

To summarize the recommendation:

- Keep the original `test_primfunc`(perhaps rename to `test_prim_func_before_lowering` to clairfy the intend)
    - This was the original intended behavior of the extractor and auto-scheduler only relied on this behavior.  
-  Add a new test `test_primfunc_lowered_with_let`
    - I understand that there is an exploration effort to try out this and possibly followup features that depends on this. Although auto-scheduler do not yet rely on this behavior, so it would be good to keep it separate.
    - Like you said it is more like an integration tests(but i agree in this case it is OK to keep it here)
-  (Optional): add a unit-test testcase with explicitly constructed tvmscript that covers the `lowered_let`. The main rationale is that the intend of the unit-test case would be clear and it becomes easier to debug in facing regression, but I feel in this case it can be optional.

Reply:
done